### ISSUE TITLE
Hook DX9,DX10,DX11 IDXGISwapChain::Present() functions cause steam FPS counter double itself

### ISSUE THREAD

**Issue by cdjzzzz** [EasyHook/EasyHook#254]
Once I use LhInstallHook() & LhSetExclusiveACL() hooking IDXGISwapChain::Present() of Dx9, Dx10 and Dx11, Steam FPS counter double itself in-game overlay showing.
Many popular games can reproduce this issue, like CSGO, Dota2, Brew-Ha, Null Vector.
You can enable Steam FPS Counter as below:
left-top Steam ->In-Game->In-game FPS counter->choose any except Off.

**Comment by spazzarama** [comment]
This is not an issue with EasyHook. You need to check the presentation count of the dx11 device as it is quite common for present to be called multiple times per frame. Not sure about dx9, but still the issue is not originating within EasyHook.

There is a sample solution in the Direct3DHook repository issues somewhere.

**Comment by spazzarama** [comment]
@cdjzzzz just to clarify, it is possible to skip EasyHook any other hooks that might have been installed AFTER EasyHook by calling the address returned by LhGetHookBypassAddress instead of calling the original method from within Present.

**Comment by cdjzzzz** [comment]
Yes, I try to hook present and render the FPS number within present hook. 
I have found the Direct3DHook repository issues you mentioned, It's a little different.

As you guess at last, I try to delay the easyhook hooking. That's to say, Steam hooks installed before EasyHook hooks. Steam FPS counter is no longer doubled.

**Comment by spazzarama** [comment]
Ok thanks @cdjzzzz. Would you mind trying with LhGetHookBypassAddress also without the delay? I think that it will also solve your issue (assuming EasyHook hook is always installed before steam one).

**Comment by cdjzzzz** [comment]
Thanks @spazzarama , the LhGetHookBypassAddress works. I see the function description says that this function must be used with great care. If I only use the return function pointer within present, any more should I care about?
BTW, I was wondering that if i sue LhGetHookBypassAddress , Am I block any other Hooks into present?

**Comment by spazzarama** [comment]
@cdjzzzz `LhGetHookBypassAddress` gives you a pointer to AFTER the EasyHook outro (thread safety etc). Therefore calling within your Present hook should not create an issue. The only time I can think would cause a problem is if you call `LhGetHookBypassAddress`, then uninstall the hook and then use the address already retrieved.

In your case this is basically what is happening:

1) original function
Instruction 1
Instruction 2
Instruction 3
..
ret

2) EasyHook hooked function (assuming the space of 1st instructions is enough for trampoline jump)
**EasyHook: jump to EasyHook trampoline address**
Instruction 2
Instruction 3
...
ret
**; EasyHook trampoline address**
EasyHook intro
**Call your hook handler function**
EasyHook outro
Instruction 1 ; relocated instructions
jump back to Instruction 2 addr

3) Steam hooked after EasyHook
**Steam: jump steam trampoline address**
Instruction 2
Instruction 3
...
ret
**; steam trampoline address**
Steam trampoline and hook logic
...
**jump to EasyHook trampoline address** ; relocated easyhook instructions

**Comment by cdjzzzz** [comment]
@spazzarama Thanks for your explaination. Now I understand what case I am encoutering.
using LhGetHookBypassAddress has some side effect as my test. Once I uninstall the hook, steam FPS overlay is also gone too. So I prefer the delay hook insatlling
